File: usememory/map/applied/map_value/sparse_matrix_multiply.scala
object Solution {
    def multiply(mat1: Array[Array[Int]], mat2: Array[Array[Int]]): Array[Array[Int]] = {
        // brute force
        // for loop over every row & column combination for O(m x k x k x n)
        
        // use memory (dynamic programming)
        // first matrix only uses rows, and second matrix only uses columns
        // we re-use each row in the first matrix n and we re-use each column in the second matrix m times
        // so we could have a map with (0, 0) -> 1, (0, 1) -> 0, (0, 2) -> 0 etc.
        // which means complexity of O(m x k + k x n)
        // however, we can do even better by only storing in the map non-zero values (we are told it is sparse)
        // and then searching for keys
        val mat1Map = mat1.zipWithIndex
                          .flatMap{ case (row, rowIndex) => 
                                row.zipWithIndex.map{case (v,colIndex) => (rowIndex, colIndex, v)}}
                          .filter{ case (_, _, v) => v != 0}
                          .groupBy{ case (x, y, v) => x }
        val mat2Map = mat2.zipWithIndex
                          .flatMap{ case (row, rowIndex) => 
                                row.zipWithIndex.map{case (v,colIndex) => (rowIndex, colIndex, v)}}
                          .filter{ case (_, _, v) => v != 0}
                          .groupBy{ case (x, y, v) => x }
        // could potentially optimise that you pick the smaller map to foldleft over 
        mat1Map.foldLeft(Array.fill[Int](mat1.length, mat2(0).length)(0)){
            case (solution, (r1, row1)) => 
                row1.foreach{
                    case (r1 : Int, c1 : Int, v1 : Int) => 
                        mat2Map.getOrElse(c1, Array[Int]()) match {
                            case row2 => row2.foreach{
                                case (r2 : Int, c2 : Int, v2 : Int) => solution(r1)(c2) = (solution(r1)(c2) + v1 * v2)
                            }
                        }
                }
                solution 
        }
    }
}

File: map/map_value/sparse_matrix_multiply.go
package data_structures

func multiply(mat1 [][]int, mat2 [][]int) [][]int {
	// brute force : given mxn and nxk matrices, complexity is m x k x n^2 = ~O(n^4), i.e.
	// each value in the final matrix has to go over n times

	// non-zero maps : we can avoid looping over the n values in a row and n values in a col
	// by storing them as maps then if on average 10% of n values are non-zero
	// complexity would be m x n + n x k + m x k x (n /10)^2 = ~O(n^2)

	// calculate cumulatively = any given entry is in final matrix is
	// r_ij = m1_i1 * m2_1j + ... m1_in * m2_nj
	// although each pair is never re-used m1_iz is re-used multiple times
	// so for m1_i1 we need m1_i1 * m2_11, m1_i1 * m2_12, ... m1_i1 * mk_1k
	// therefore we can have a map from mat2 where 1 -> [m2_11, m2_12, ... mk_1k]
	// but we only store non-zero items
	// complexity, build the mat2 map O(nxk)
	// we could build a map of mat1 - but no point as would have to loop over that
	// loop over mat1 O(mxn) and for each value we need to look up all values for that
	// column i.e. O(n), but only need to do loop for non-zero values which if we assume
	// is only 10% of the time which saves k iterations
	// O(nxk + mxn + mxn/10 * n/10) = ~O(n^2)

	// there are two ways we can build up the final mxk matrix
	// 1. calculate each mxk entry which uses mth row and kth column, although you will
	//    have to look up each component multiple times
	// 2. calculate all the values used by a component summing incrementally
	mLen := len(mat1)
	nLen := len(mat2)
	kLen := len(mat2[0])
	mat2Map := make(map[int]map[int]int) // nxk: n row -> [k col -> value]
	// create mat2 map O(nxk)
	for n := 0; n < nLen; n++ {
		// nxk matrix
		for k := 0; k < kLen; k++ {
			if mat2[n][k] != 0 {
				if map2KCol, found := mat2Map[n]; found {
					map2KCol[k] = mat2[n][k]
					mat2Map[n] = map2KCol
				} else {
					newMap2KCol := make(map[int]int)
					newMap2KCol[k] = mat2[n][k]
					mat2Map[n] = newMap2KCol
				}
			}
		}
	}
	result := make([][]int, mLen)
	// initialize matrix O(mxk)
	for m := range result {
		result[m] = make([]int, kLen)
		for k := range result[m] {
			result[m][k] = 0
		}
	}
	// add non-zero mat1[m][n] everywhere it is used O(mxnx(k/10))
	for m := 0; m < mLen; m++ {
		for n := 0; n < nLen; n++ {
			if mat1[m][n] != 0 {
				// multiply ith row of mat1 by ith col of mat2
				if mat2NRow, ok := mat2Map[n]; ok {
					for kCol, mat2NK := range mat2NRow {
						result[m][kCol] += mat1[m][n] * mat2NK
					}
				}
			}
		}
	}

	return result
}
